src/rely/Mock.re

```diff
@@ -0,0 +1,161 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */;
+
+type result('a) =
+  | Return('a)
+  | Exception(exn);
+
+type t('a, 'b, 'c) = {
+  calls: ref(list('c)),
+  results: ref(list(result('b))),
+  fn: 'a,
+  originalImplementation: 'a,
+  implementation: ref('a),
+};
+
+let fn = mock => mock.fn;
+let getCalls = mock => mock.calls^;
+let getResults = mock => mock.results^;
+let wasCalled = mock =>
+  switch (mock.calls^) {
+  | [] => false
+  | [first, ...rest] => true
+  };
+let resetHistory = mock => {
+  mock.calls := [];
+  mock.results := [];
+};
+let resetImplementation = mock =>
+  mock.implementation := mock.originalImplementation;
+let reset = mock => {
+  resetHistory(mock);
+  resetImplementation(mock);
+};
+let changeImplementation = (fn, mock) => mock.implementation := fn;
+
+let mock1 = fn => {
+  let calls = ref([]);
+  let results = ref([]);
+  let implementation = ref(fn);
+  let wrappedFn = arg => {
+    calls := calls^ @ [arg];
+    switch (implementation^(arg)) {
+    | returnValue =>
+      results := results^ @ [Return(returnValue)];
```

Comment:
This concatenation is slow, and will progressively slow down the execution. You want to prepend to the list for an O(1) operation and maintain the list of results in reverse order. (Same for calls)